### ISSUE TITLE
Screen readers, ARIA, and native tables

### ISSUE THREAD

**Issue by artlawry** [fromtheoutfit/responsive_faux_table#2]
You may want to see how screen readers handle definition lists and tables to make sure the experience remains the same for the visually impaired.

If they differ, ARIA might have roles that will assist in recreating the tabular data experience.

Alternately, you could simply try doing the same thing with a table and modifying the display properties of td, tr, tbody, etc. to achieve something similar.

**Comment by marcamos** [comment]
All good ideas, sir! When time allows, I'll give them a shot.

**Comment by artlawry** [comment]
Looks like "grid", "row", and "grid-cell" might accomplish this, but I'm not well-versed in ARIA, nor have I seen how screen-readers treat ARIA grid content vs natural table content.

**Comment by marcamos** [comment]
I just tinkered with this for a couple of hours, and published a separate branch called _better-accessibility_; see [this new branch's index.html page](https://github.com/fromtheoutfit/responsive_faux_table/blob/better-accessibility/index.html)—specifically the `role` and `aria-*` attributes used throughout.

I think it was, largely, a failure. I'm very inexperienced with all things ARIA, and doubly-so with OS X's VoiceOver app (which is what I used to _hear_ the results of my efforts), along with [this simple guide](http://www.apple.com/voiceover/info/guide/_1131.html).

The end-result: VoiceOver reads through the first row, successfully saying aloud the key/value pairs, but it fails to go to the second row and beyond. This could be because I don't know _how_ get control VoiceOver… but I suspect it's because we're trying to do something that just doesn't work.

I think it might be wasted effort on my part; I don't know how to use the tool, and as a result I cannot know if the code is right, wrong, or somewhere in between.

I'm happy to leave this issue _open_ in case somebody comes along and schools me on what's right/wrong/etc.

**Comment by artlawry** [comment]
I'm not well versed in ARIA either, so I understand. If I had to guess it has something to do with the structure of a definition list and the structure of a table not being compatible.

In your first row, you essentially have a `<th>` immediately followed by a `<td>` when those are typically in separate rows. this essentially makes your first row have twice as many cells as the next, and this may be where VoiceOver is choking.

I can't see a way to immediately make this work, except perhaps to add a structure before the first DL that just contains the column headings and to never mark up the `<dt>`s as column headings. This still might not work if ARIA requires every child of `role='row'` to be a `role='columnheader'` or a `role='gridcell'`.

I would try removing all roles and seeing how VoiceOver behaves, then, if all your content is read correctly, try the extra container of headers. Also, remember that if content is `display: none;` readers may not speak it, but `position: absolute; left: -10000px;` and `overflow: hidden;` on the parent should leave it audible to the visually impaired.
